### ISSUE TITLE
dotnet-svcutil: ability to generate data members with EmitDefaultValue=false and no IsRequired=true

### ISSUE THREAD

**Issue by abelykh0** [dotnet/wcf#3961]
**Describe the bug**
I need to be able to generate a data members with EmitDefaultValue=false and no IsRequired=true. 

**To Reproduce**
Use wsdl like this:

```
<?xml version="1.0" encoding="utf-8"?>
<w:definitions xmlns:http="http://schemas.xmlsoap.org/wsdl/http/" xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/" xmlns:mime="http://schemas.xmlsoap.org/wsdl/mime/" xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/" xmlns:tm="http://microsoft.com/wsdl/mime/textMatching/" xmlns:s="http://www.w3.org/2001/XMLSchema" xmlns:s0="urn:schemas-microsoft-com:xml-analysis" targetNamespace="urn:schemas-microsoft-com:xml-analysis" xmlns:w="http://schemas.xmlsoap.org/wsdl/">
  <w:types>
    <s:schema attributeFormDefault="qualified" elementFormDefault="qualified" targetNamespace="urn:schemas-microsoft-com:xml-analysis">
      <s:element name="Discover">
        <s:complexType>
          <s:sequence>
            <s:element name="RequestType" type="s:string" />
            <s:element name="Restrictions">
              <s:complexType>
                <s:sequence>
                  <s:element name="RestrictionList">
                    <s:complexType>
                      <s:sequence>
                        <s:element name="CATALOG_NAME">
                          <s:simpleType>
                            <s:restriction base="s:string">
                              <s:maxLength value="32" />
                            </s:restriction>
                          </s:simpleType>
                        </s:element>
                        <s:element name="CUBE_NAME">
                          <s:simpleType>
                            <s:restriction base="s:string">
                              <s:maxLength value="70" />
                            </s:restriction>
                          </s:simpleType>
                        </s:element>
                        <s:element name="CUBE_TYPE">
                          <s:simpleType>
                            <s:restriction base="s:string">
                              <s:maxLength value="70" />
                            </s:restriction>
                          </s:simpleType>
                        </s:element>
                        <s:element name="DIMENSION_UNIQUE_NAME">
                          <s:simpleType>
                            <s:restriction base="s:string">
                              <s:maxLength value="32" />
                            </s:restriction>
                          </s:simpleType>
                        </s:element>
                        <s:element name="HIERARCHY_UNIQUE_NAME">
                          <s:simpleType>
                            <s:restriction base="s:string">
                              <s:maxLength value="65" />
                            </s:restriction>
                          </s:simpleType>
                        </s:element>
                        <s:element name="LEVEL_UNIQUE_NAME">
                          <s:simpleType>
                            <s:restriction base="s:string">
                              <s:maxLength value="108" />
                            </s:restriction>
                          </s:simpleType>
                        </s:element>
                        <s:element name="LEVEL_NUMBER" type="s:unsignedInt" />
                        <s:element name="MEMBER_UNIQUE_NAME">
                          <s:simpleType>
                            <s:restriction base="s:string">
                              <s:maxLength value="143" />
                            </s:restriction>
                          </s:simpleType>
                        </s:element>
                        <s:element name="TREE_OP" type="s:unsignedInt" />
                        <s:element name="MEASURE_UNIQUE_NAME">
                          <s:simpleType>
                            <s:restriction base="s:string">
                              <s:maxLength value="143" />
                            </s:restriction>
                          </s:simpleType>
                        </s:element>
                        <s:element name="PROPERTY_NAME">
                          <s:simpleType>
                            <s:restriction base="s:string">
                              <s:maxLength value="30" />
                            </s:restriction>
                          </s:simpleType>
                        </s:element>
                      </s:sequence>
                    </s:complexType>
                  </s:element>
                </s:sequence>
              </s:complexType>
            </s:element>
            <s:element name="Properties">
              <s:complexType>
                <s:complexContent mixed="false">
                  <s:extension base="s0:PropertiesType">
                    <s:sequence>
                      <s:element name="PropertyList">
                        <s:complexType>
                          <s:sequence>
                            <s:element name="AxisFormat" type="s:string" />
                            <s:element name="BeginRange">
                              <s:simpleType>
                                <s:restriction base="s:int">
                                  <s:totalDigits value="10" />
                                </s:restriction>
                              </s:simpleType>
                            </s:element>
                            <s:element name="EndRange">
                              <s:simpleType>
                                <s:restriction base="s:int">
                                  <s:totalDigits value="10" />
                                </s:restriction>
                              </s:simpleType>
                            </s:element>
                            <s:element name="Catalog" type="s:string" />
                            <s:element name="Cube" type="s:string" />
                            <s:element name="Content" type="s:string" />
                            <s:element name="DataSourceInfo" type="s:string" />
                            <s:element name="Format" type="s:string" />
                            <s:element name="LocalIdentifier" type="s:string" />
                            <s:element name="MDXSupport" type="s:string" />
                            <s:element name="ProviderName" type="s:string" />
                            <s:element name="ProviderVersion" type="s:string" />
                            <s:element name="StateSupport" type="s:string" />
                            <s:element name="Timeout">
                              <s:simpleType>
                                <s:restriction base="s:int">
                                  <s:totalDigits value="10" />
                                </s:restriction>
                              </s:simpleType>
                            </s:element>
                          </s:sequence>
                        </s:complexType>
                      </s:element>
                    </s:sequence>
                  </s:extension>
                </s:complexContent>
              </s:complexType>
            </s:element>
          </s:sequence>
        </s:complexType>
      </s:element>
      <s:element name="DiscoverResponse">
        <s:complexType>
          <s:sequence>
            <s:element minOccurs="0" name="return">
              <s:complexType>
                <s:sequence>
                  <s:any />
                </s:sequence>
              </s:complexType>
            </s:element>
          </s:sequence>
        </s:complexType>
      </s:element>
      <s:complexType name="PropertiesType" />
    </s:schema>
  </w:types>
  <w:message name="DiscoverSoapIn">
    <w:part name="parameters" element="s0:Discover" />
  </w:message>
  <w:message name="DiscoverSoapOut">
    <w:part name="parameters" element="s0:DiscoverResponse" />
  </w:message>
  <w:portType name="MsXmlAnalysisSoap">
    <w:operation name="Discover">
      <w:input message="s0:DiscoverSoapIn" />
      <w:output message="s0:DiscoverSoapOut" />
    </w:operation>
  </w:portType>
  <w:binding name="MsXmlAnalysisSoap" type="s0:MsXmlAnalysisSoap">
    <soap:binding transport="http://schemas.xmlsoap.org/soap/http" />
    <w:operation name="Discover">
      <soap:operation soapAction="urn:schemas-microsoft-com:xml-analysis:Discover" style="document" />
      <w:input>
        <soap:body use="literal" />
      </w:input>
      <w:output>
        <soap:body use="literal" />
      </w:output>
    </w:operation>
  </w:binding>
  <w:service name="MsXmlAnalysis">
    <w:port name="MsXmlAnalysisSoap" binding="s0:MsXmlAnalysisSoap">
      <soap:address location="http://demo:8000/sap/bw/xml/soap/xmla" />
    </w:port>
  </w:service>
</w:definitions>
```

**Expected behavior**
Only EmitDefaultValue=false and no IsRequired=true

**Additional context**
Unfortunately, 
```
[System.Runtime.Serialization.DataMemberAttribute(IsRequired=true, EmitDefaultValue=false)]
```
causes the following exception when trying to connect to Essbase:

SerializationException: Member CUBE_NAME in type MsXmlAnalysisSoap.Restrictions+RestrictionListType cannot be serialized. This exception is usually caused by trying to use a null value where a null value is not allowed. The 'CUBE_NAME' member is set to its default value (usually null or zero). The member's EmitDefault setting is 'false', indicating that the member should not be serialized. However, the member's IsRequired setting is 'true', indicating that it must be serialized. This conflict cannot be resolved.  Consider setting 'CUBE_NAME' to a non-default value. Alternatively, you can change the EmitDefaultValue property on the DataMemberAttribute attribute to true, or changing the IsRequired property to false.

**Comment by StephenBonikowsky** [comment]
@abelykh0 thanks for reporting the problem, we'll have someone investigate and get back to you.

**Comment by dasetser** [comment]
@abelykh0 This looks like it's by design. The wsdl you posted defines CUBE_NAME as:
```XML
<s:element name="CUBE_NAME">
  <s:simpleType>
    <s:restriction base="s:string">
      <s:maxLength value="70" />
    </s:restriction>
  </s:simpleType>
</s:element>
```
The element in the wsdl that corresponds to IsRequired is minOccurs. If minOccurs > 0 then IsRequired should be true. Since this element doesn't specify minOccurs, it defaults to 1. This means the wsdl is saying IsRequired needs to be true.

EmitDefaultValue is getting set to false because it has to be. CUBE_NAME is based on the string class, so the default would be null (or nil in XML). Since the wsdl doesn't specify nillable=true, it defaults to false. So we can't set the value to nil, but the default for the type should be nil. This means the only option is to not emit the default value.

In short, your wsdl is saying the element is required (by not specifying minOccurs = 0), and is saying the element can't be null (by not setting nillable = true). So the generated code is correct. To avoid the exception you're getting you should ensure this element always has a value other than null (string.Empty at least if you don't have an actual value). The server is expecting this to always have a value.

**Comment by abelykh0** [comment]
OK, great. What about LEVEL_NUMBER? Is it possible to generate data member for it with EmitDefaultValue=false?

**Comment by dasetser** [comment]
The generated code for LEVEL_NUMBER doesn't include EmitDefaultValue=false because it's only recommended to set this if you need to. This variable is an unsigned int, which is a value type. The fact that it's not nillable is okay because we can use 0 as the default value instead. 

If you need to set this for some reason you can manually change it on the client side, or if you own the service side you can set it on the DataMember with `[DataMember(EmitDefaultValue = false)]`, which will modify the wsdl to tell the tool to generate it with that attribute.
